Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -350,64 +355,135 @@ lemma hasDerivAt_binaryEntropy {x : ℝ} (xne0: x ≠ 0) (gne1 : x ≠ 1) :
   exact binaryEntropy_eq
   exact (deriv_binaryEntropy' xne0 gne1).symm
 
+lemma hasDerivAt_qaryEntropy {q : ℕ} {x : ℝ} (qnot1 : q ≠ 1) (xne0: x ≠ 0) (gne1 : x ≠ 1) :
+    HasDerivAt (qaryEntropy q) (log (q - 1) + log (1 - x) - log x) x := by
+  have diffAt :
+      DifferentiableAt ℝ (fun p => p * log (q - 1) - p * log p - (1 - p) * log (1 - p)) x := by
+    apply DifferentiableAt.sub
+    apply DifferentiableAt.sub
+    apply DifferentiableAt.mul
+    exact differentiableAt_id'
+    apply DifferentiableAt.log
+    simp only [ne_eq, differentiableAt_const]
+    exact sub_ne_zero_of_ne (Nat.cast_ne_one.mpr qnot1)
+    apply DifferentiableAt.mul
+    exact differentiableAt_id'
+    apply DifferentiableAt.log differentiableAt_id' xne0
+    apply DifferentiableAt.mul
+    apply DifferentiableAt.sub
+    apply differentiableAt_const
+    exact differentiableAt_id'
+    exact differentiableAt_log_const_neg gne1
+  convert hasDerivAt_deriv_iff.mpr diffAt using 1
+  exact Eq.symm (deriv_qaryEntropy' xne0 gne1)
```

Comment:
Ouch. Doesn't `fun_prop` work here?

Reply:
Simply `by fun_prop` does not work. lecopivo already pointed out some ways to shorten things using `fun_prop` but I haven't spent enough time on it to be able to do it yet. In any case, right now my focus is getting the definitions and statements right. I'll try to make the proofs shorter next.